### PR TITLE
e2e: ICS27 interchain accounts x/group integration

### PR DESCRIPTION
**Pull request by damiannolan**
## Description

- Adding e2e with `x/group` integration for interchain-accounts.

closes: #2039 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: e2e/tests/interchain_accounts/groups_test.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	t.Run("fund interchain account wallet", func(t *testing.T) {
+		err := chainB.SendFunds(ctx, ibctest.FaucetAccountKeyName, ibc.WalletAmount{
+			Address: interchainAccAddr,
+			Amount:  testvalues.StartingTokenAmount,
+			Denom:   chainB.Config().Denom,
+		})
+		s.Require().NoError(err)
+	})
+
+	t.Run("submit proposal submit tx", func(t *testing.T) {
```
Comment (crodriguezvega) [inline e2e/tests/interchain_accounts/groups_test.go:130]:
```suggestion
	t.Run("submit proposal for submit tx", func(t *testing.T) {
```

File: e2e/testconfig/testconfig.go [hunk 2]
```diff
@@ -197,6 +199,16 @@ func defaultModifyGenesis(denom string) func([]byte) ([]byte, error) {
...
 
+		icaGenesisState := icagenesistypes.DefaultGenesis()
+		icaGenesisState.HostGenesisState.Params.AllowMessages = []string{"*"}
+
+		icaGenBz, err := cdc.MarshalJSON(icaGenesisState)
+		if err != nil {
+			return nil, fmt.Errorf("failed to marshal interchain accounts genesis state: %w", err)
+		}
+
+		appState[icatypes.ModuleName] = icaGenBz
```
Comment (damiannolan) [inline e2e/testconfig/testconfig.go:210]: I guess this is not required anymore. ref: #2290

File: e2e/tests/interchain_accounts/groups_test.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+}
+
+type InterchainAccountsGroupsTestSuite struct {
+	testsuite.E2ETestSuite
+}
+
+func (s *InterchainAccountsGroupsTestSuite) QueryGroupPolicyAddress(ctx context.Context, chain ibc.Chain) string {
+	queryClient := s.GetChainGRCPClients(chain).GroupsQueryClient
+	res, err := queryClient.GroupPoliciesByGroup(ctx, &grouptypes.QueryGroupPoliciesByGroupRequest{
+		GroupId: 1, // always use the initial group id
```
Comment (damiannolan) [inline e2e/tests/interchain_accounts/groups_test.go:37]: extract to const

File: e2e/tests/interchain_accounts/groups_test.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	chainAAddress := chainAWallet.Bech32Address(chainA.Config().Bech32Prefix)
+
+	chainBWallet := s.CreateUserOnChainB(ctx, testvalues.StartingTokenAmount)
+	chainBAddress := chainBWallet.Bech32Address(chainB.Config().Bech32Prefix)
+
+	t.Run("create group with new threshold decision policy", func(t *testing.T) {
+		members := []grouptypes.MemberRequest{
+			{
+				Address: chainAAddress,
+				Weight:  "1",
```
Comment (damiannolan) [inline e2e/tests/interchain_accounts/groups_test.go:69]: extract to const?

File: e2e/testsuite/testsuite.go [hunk 5]
```diff
@@ -224,6 +227,13 @@ func (s *E2ETestSuite) GetChains(chainOpts ...testconfig.ChainOptionConfiguratio
 // Once the broadcast response is returned, we wait for a few blocks to be created on both chain A and chain B.
 func (s *E2ETestSuite) BroadcastMessages(ctx context.Context, chain *cosmos.CosmosChain, user *ibc.Wallet, msgs ...sdk.Msg) (sdk.TxResponse, error) {
 	broadcaster := cosmos.NewBroadcaster(s.T(), chain)
+
+	configureGasFactoryOpt := func(factory tx.Factory) tx.Factory {
+		return factory.WithGas(500000)
+	}
+
+	broadcaster.ConfigureFactoryOptions(configureGasFactoryOpt)
```
Comment (damiannolan) [inline e2e/testsuite/testsuite.go:235]: This was required to successfully exec group proposals

File: e2e/go.mod [hunk 6]
```diff
@@ -5,16 +5,16 @@ go 1.18
 replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
 require (
-	github.com/cosmos/cosmos-sdk v0.46.0
-	github.com/cosmos/ibc-go/v5 v5.0.0-beta1
+	github.com/cosmos/cosmos-sdk v0.46.1
+	github.com/cosmos/ibc-go/v6 v6.0.0-20220923210821-94e61c0f56ac
```
Comment (damiannolan) [inline e2e/go.mod:9]: Ideally we can tag a `v6-alpha1` and start the e2e grind.

**Comment by chatton** [comment]
All the e2e tests will fail as the [ibc tests ibc-go version](https://github.com/strangelove-ventures/ibctest/blob/main/go.mod#L13) needs to match the one in our `e2e` go mod. This is an sdk limitation at the moment as duplicate types are registered.

File: e2e/testconfig/testconfig.go (hunk 2 above)
Comment (colin-axner) [reply e2e/testconfig/testconfig.go:210]: Yea lets just remove and test this functionality against >= v6

File: e2e/testsuite/testsuite.go (hunk 5 above)
Comment (colin-axner) [reply e2e/testsuite/testsuite.go:235]: should we make the gas argument a global const?

File: e2e/testsuite/grpc_query.go [hunk 7]
```diff
@@ -49,14 +49,14 @@ func (s *E2ETestSuite) QueryPacketCommitment(ctx context.Context, chain ibc.Chai
 // QueryInterchainAccount queries the interchain account for the given owner and connectionId.
 func (s *E2ETestSuite) QueryInterchainAccount(ctx context.Context, chain ibc.Chain, owner, connectionId string) (string, error) {
 	queryClient := s.GetChainGRCPClients(chain).ICAQueryClient
-	res, err := queryClient.InterchainAccount(ctx, &intertxtypes.QueryInterchainAccountRequest{
+	res, err := queryClient.InterchainAccount(ctx, &controllertypes.QueryInterchainAccountRequest{
```
Comment (colin-axner) [inline e2e/testsuite/grpc_query.go:52]: will this break cross compatibility tests?

File: e2e/tests/interchain_accounts/groups_test.go (hunk 4 above)
Comment (colin-axner) [reply e2e/tests/interchain_accounts/groups_test.go:69]: probably a good habit

Comment (colin-axner) [reply e2e/tests/interchain_accounts/groups_test.go:69]: what is a threshold decision policy? What does the weight do?

File: e2e/tests/interchain_accounts/groups_test.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+			},
+		}
+
+		decisionPolicy := grouptypes.NewThresholdDecisionPolicy("1", time.Duration(time.Minute), time.Duration(0))
```
Comment (colin-axner) [inline e2e/tests/interchain_accounts/groups_test.go:73]: What are the time values doing?

File: e2e/tests/interchain_accounts/groups_test.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+		msgBankSend := &banktypes.MsgSend{
+			FromAddress: interchainAccAddr,
+			ToAddress:   chainBAddress,
+			Amount:      sdk.NewCoins(testvalues.DefaultTransferAmount(chainB.Config().Denom)),
+		}
+
+		cfg := simappparams.MakeTestEncodingConfig()
+		banktypes.RegisterInterfaces(cfg.InterfaceRegistry)
+		cdc := codec.NewProtoCodec(cfg.InterfaceRegistry)
+
```
Comment (colin-axner) [inline e2e/tests/interchain_accounts/groups_test.go:140]: We should add a cdc to our testsuite in a followup pr

File: e2e/tests/interchain_accounts/groups_test.go [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+		bz, err := icatypes.SerializeCosmosTx(cdc, []sdk.Msg{msgBankSend})
+		s.Require().NoError(err)
+
+		packetData := icatypes.InterchainAccountPacketData{
+			Type: icatypes.EXECUTE_TX,
+			Data: bz,
+			Memo: "e2e",
+		}
+
+		timeoutTimestamp := time.Now().Add(time.Hour * 24).UnixNano() // TODO: find a better solution
```
Comment (colin-axner) [inline e2e/tests/interchain_accounts/groups_test.go:150]: This is relative now, just use 1 minute or something

**Review by colin-axner — COMMENTED** [review]
Woot woot!! Superb work! Left some suggestions, will review in detail once it is updated to an alpha tag

File: e2e/testsuite/grpc_query.go (hunk 7 above)
Comment (damiannolan) [reply e2e/testsuite/grpc_query.go:53]: Yeah, great catch! 

I've re-added the intertx query client and a `QueryInterchainAccountLegacy` method which is used by the existing tests

File: e2e/tests/interchain_accounts/groups_test.go (hunk 9 above)
Comment (damiannolan) [reply e2e/tests/interchain_accounts/groups_test.go:170]: Definitely agree. https://github.com/cosmos/ibc-go/issues/2512

File: e2e/tests/interchain_accounts/groups_test.go (hunk 4 above)
Comment (damiannolan) [reply e2e/tests/interchain_accounts/groups_test.go]: The `ThresholdDecisionPolicy` defines a weighted threshold that x amount of YES votes must meet in order to succeed. Each group member can carry a different weight which is accounted for in the tally. 

I've moved a bunch of fields to consts at the top of file and added godocs (some copy pasted from sdk).

File: e2e/testsuite/testsuite.go (hunk 5 above)
Comment (damiannolan) [reply e2e/testsuite/testsuite.go]: Done

File: e2e/tests/interchain_accounts/groups_test.go (hunk 8 above)
Comment (damiannolan) [reply e2e/tests/interchain_accounts/groups_test.go]: see consts added to top of file

**Review by crodriguezvega — APPROVED** [review]
LGTM. Thanks, @damiannolan!

File: e2e/testsuite/testsuite.go [hunk 11]
```diff
@@ -365,8 +379,10 @@ func (s *E2ETestSuite) initGRPCClients(chain *cosmos.CosmosChain) {
 		ClientQueryClient:  clienttypes.NewQueryClient(grpcConn),
 		ChannelQueryClient: channeltypes.NewQueryClient(grpcConn),
 		FeeQueryClient:     feetypes.NewQueryClient(grpcConn),
-		ICAQueryClient:     intertxtypes.NewQueryClient(grpcConn),
+		ICAQueryClient:     controllertypes.NewQueryClient(grpcConn),
+		InterTxQueryClient: intertxtypes.NewQueryClient(grpcConn),
```
Comment (colin-axner) [inline e2e/testsuite/testsuite.go:383]: Maybe we should open up an issue to remove all the intertx stuff once v5 is no longer supported?

File: e2e/tests/interchain_accounts/groups_test.go [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+		s.Require().NoError(err)
+		s.AssertValidTxResponse(txResp)
+	})
+
+	t.Run("start relayer", func(t *testing.T) {
+		s.StartRelayer(relayer)
+	})
+
+	t.Run("verify interchain account registration success", func(t *testing.T) {
+		interchainAccAddr, err = s.QueryInterchainAccount(ctx, chainA, groupPolicyAddr, ibctesting.FirstConnectionID)
```
Comment (colin-axner) [inline e2e/tests/interchain_accounts/groups_test.go:143]: I guess you could assert the addr is non empty

File: e2e/tests/interchain_accounts/groups_test.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+		s.Require().NoError(err)
+
+		channels, err := relayer.GetChannels(ctx, s.GetRelayerExecReporter(), chainA.Config().ChainID)
+		s.Require().NoError(err)
+		s.Require().Equal(len(channels), 2)
```
Comment (colin-axner) [inline e2e/tests/interchain_accounts/groups_test.go:148]: 1 transfer channel, 1 ica channel?

File: e2e/tests/interchain_accounts/groups_test.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+		s.AssertValidTxResponse(txResp)
+	})
+
+	t.Run("verify tokens transferred", func(t *testing.T) {
+		s.Require().NoError(test.WaitForBlocks(ctx, 10, chainA, chainB), "failed to wait for blocks")
+
+		balance, err := chainB.GetBalance(ctx, chainBAddress, chainB.Config().Denom)
+		s.Require().NoError(err)
+
+		_, err = chainB.GetBalance(ctx, interchainAccAddr, chainB.Config().Denom)
```
Comment (colin-axner) [inline e2e/tests/interchain_accounts/groups_test.go:208]: should you check that this balance has be subtracted as a sanity check?

**Review by colin-axner — APPROVED** [review]
Excellent! :sunglasses: Super cool to see a test for this. Thanks for handling this work!

File: e2e/testsuite/testsuite.go (hunk 11 above)
Comment (damiannolan) [reply e2e/testsuite/testsuite.go:383]: Sure, I can open an issue. When is v5 no longer supported? Like 1 year from now?

File: e2e/tests/interchain_accounts/groups_test.go (hunk 12 above)
Comment (damiannolan) [reply e2e/tests/interchain_accounts/groups_test.go:143]: Added as an extra sanity check but I don't think its reachable. Something would be wrong with the query then I think

File: e2e/tests/interchain_accounts/groups_test.go (hunk 13 above)
Comment (damiannolan) [reply e2e/tests/interchain_accounts/groups_test.go]: Yup! Added inline comment 👍

File: e2e/tests/interchain_accounts/groups_test.go (hunk 14 above)
Comment (damiannolan) [reply e2e/tests/interchain_accounts/groups_test.go]: Done :)

**Pull request merged**